src/software/multithreading/threaded_observer.h

```diff
@@ -5,25 +5,35 @@
 
 #include "software/multithreading/observer.h"
 
+enum class ThreadedObserverOrdering
+{
+    QUEUE,  // first in, first out
+    STACK   // last in, first out
+};
+
 /**
  * The general usage of this class should be to extend it, then override
  * `onValueReceived` with whatever custom functionality should occur when a new value
  * is received.
  *
  * @tparam T The type of object this class is observing
+ * @tparam Ordering the ordering of which objects in the internal buffer are iterated
+ *                  through when calling `onValueReceived`.
  */
-template <typename T>
-class ThreadedObserver : public Observer<T>
+template <typename T, ThreadedObserverOrdering Ordering>
+class GenericThreadedObserver : public Observer<T>
```

Comment:
So I think we can avoid using templating here, which would be a win in terms of comprehensibility (templating is hard to understand, especially for newer members).

Could we:
- Leave `ThreadedObserver<T>` as-is, with an additional `std::optional<T> getNextValue()` which we use in `continuouslyPullValuesFromBuffer`
- Create `FirstInFirstOutThreadedObserver<T>` and `LastInFirstOutThreadedObserver<T>` that extend `ThreadedObserver<T>` and just override `getNextValue()`?